### PR TITLE
Max num comps

### PR DESCRIPTION
**Pull request by claytonstrawn**
When fitting very complex spectra, such as with impact parameters < 10 kpc, some students of mine would occasionally have significant components that would not be detected. We eventually decided it was because we had "run out of components" and after some digging we found that this was because there was a maximum of 8 components per complex. 

While normally this has no serious impact, and generally using more than 8 components is running against the error limit of the fitter, it was frustrating to not be able to explicitly see and mess with this parameter. I added an optional argument `maxNumComps` to `generate_total_fit` to allow this to be changed by the user.

Here's some examples below, showing this does come into play with different numbers. 8 is a perfectly good default, but I would prefer if it wasn't hard-coded.

mnc5
mnc10
mnc20

In looking into the source code and comparing with the documentation, I found what might be a more important problem also. The docs say [https://trident.readthedocs.io/en/latest/absorption_spectrum_fit.html](url)

> There are several other conditions under which the cycle of adding and optimizing lines will halt. If the error of the optimized fit from adding a line is an order of magnitude worse than the error of the fit without that line, then it is assumed that the fitting has become unstable and the latest line is removed. **Lines are also prevented from being added if the total number of lines is greater than the number of elements in the flux array being fit divided by 3**. This is because there must not be more free parameters in a fit than the number of points to constrain them. (emphasis mine)

However, in line 355, instead of breaking
`if np.size(linesP)*3>=len(x):`
it said to break
`if np.size(linesP)+3>=len(x):`

Which I think was just a typo. 

PS: I would use the normal style guide and name this new argument `max_num_comps`, but the other arguments in this file are in camelcase so I followed that. Maybe the whole file needs to be restyled?

### CODE REVIEW

File: trident/absorption_spectrum/absorption_spectrum_fit.py [hunk 1]
```diff
@@ -80,7 +81,11 @@ def generate_total_fit(x, fluxData, orderFits, speciesDicts,
         if attempting to split a region for being larger than maxlength
         the point of the split must have a flux greater than splitLim
         (ie: absorption greater than splitLim). Default= .99.
+    
+    :maxNumComps: int, optional
 
+        maximum number of voight profiles to fit at once. Default is 8 total components.
```
Comment (chummels) [inline trident/absorption_spectrum/absorption_spectrum_fit.py:87]:
```suggestion
        maximum number of voigt profiles to fit at once. Default is 8 total components.
```

File: trident/absorption_spectrum/absorption_spectrum_fit.py [hunk 2]
```diff
@@ -248,7 +253,13 @@ def _complex_fit(x, yDat, yFit, initz, minSize, errBound, speciesDict,
         dictionary containing all relevant parameters needed
         to create an absorption line of a given species (f,Gamma,lambda0)
         as well as max and min values for parameters to be fit
+    
+    :maxNumComps: int
 
+        maximum number of voight profiles to fit at once. If the loop is
```
Comment (chummels) [inline trident/absorption_spectrum/absorption_spectrum_fit.py:259]:
```suggestion
        maximum number of voigt profiles to fit at once. If the loop is
```

File: trident/absorption_spectrum/absorption_spectrum_fit.py [hunk 3]
```diff
@@ -341,7 +352,7 @@ def _complex_fit(x, yDat, yFit, initz, minSize, errBound, speciesDict,
                 break
 
         #If too many lines
-        if np.shape(linesP)[0]>8 or np.size(linesP)+3>=len(x):
+        if np.shape(linesP)[0]>maxNumComps or np.size(linesP)*3>=len(x):
```
Comment (chummels) [inline trident/absorption_spectrum/absorption_spectrum_fit.py:355]: Good catch!

**Review by chummels — APPROVED** [review]
Hey @claytonstrawn !  Thanks for the review.  This part of the code was added by @eganhila , so I'm not so familiar with it.  But I think your changes look good, both in the bug catch as well as adding the functionality to specify the number of components to fit.  As far as the style goes, you're right that the style of this code is different from the rest of the trident source, primarily because it was written independent of the rest of trident.  Your choice of style here is fine since it matches the existing format.  I just included a couple of spelling corrections but beyond that, I think this is good to go.

File: trident/absorption_spectrum/absorption_spectrum_fit.py (hunk 2 above)
Comment (claytonstrawn) [reply trident/absorption_spectrum/absorption_spectrum_fit.py]: It's crazy that I still get this wrong so often. Thanks for catching it!

**Pull request merged**